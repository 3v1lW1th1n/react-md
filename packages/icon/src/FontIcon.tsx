import React, {
  forwardRef,
  FunctionComponent,
  HTMLAttributes,
  ReactNode,
  CSSProperties,
} from "react";
import cn from "classnames";
import { IWithForwardedRef } from "@react-md/utils";

const ICON_SIZE = 24;
const DENSE_ICON_SIZE = 20;

export interface IFontIconProps extends HTMLAttributes<HTMLElement> {
  /**
   * The font icon class name to use.
   */
  iconClassName?: string;

  /**
   * Boolean if the font icon should use the dense spec.
   */
  dense?: boolean;

  /**
   * Any children to render to create the font icon. This is required for material-icons.
   */
  children?: ReactNode;

  /**
   * Either a boolean that will enforce the 24x24 size of the font icon or a number of the size
   * to enforce. This is useful when using other font icon libraries that do not have a consistent
   * size.
   */
  forceSize?: boolean | number;

  /**
   * Boolean if the `forceSize` prop should also force the `font-size` instead of only `width` and
   * `height`.
   */
  forceFontSize?: boolean;
}

export interface IFontIconDefaultProps {
  dense: boolean;
  iconClassName: string;
  forceSize: boolean;
  forceFontSize: boolean;
}

export type FontIconWithDefaultProps = IFontIconProps &
  IFontIconDefaultProps &
  IWithForwardedRef;

/**
 * A utility function that will merge the different inline styles together for the `FontIcon` if
 * any of the resizing props are enabled.
 */
function createStyles(
  style: CSSProperties | undefined,
  forceSize: boolean | number,
  forceFontSize: boolean,
  dense: boolean
) {
  let nextStyle = style;
  if (typeof forceSize === "boolean" && forceSize) {
    const size = dense ? DENSE_ICON_SIZE : ICON_SIZE;
    nextStyle = {
      fontSize: forceFontSize ? size : undefined,
      height: size,
      width: size,
      ...style,
    };
  } else if (typeof forceSize === "number") {
    nextStyle = {
      fontSize: forceFontSize ? forceSize : undefined,
      height: forceSize,
      width: forceSize,
      ...style,
    };
  }

  return nextStyle;
}

/**
 * The `FontIcon` component is used for rendering a font-icon library's
 * icon. The default is to use the `material-icons` library, but others
 * can be used as well.
 *
 * If you are using another font icon library that does not always create icons with
 * a perfect 1:1 scale (such as font awesome), it is recommended to use the `forceSize`
 * and `forceFontSize` props to fix the sizing issues.
 */
const FontIcon: FunctionComponent<
  IFontIconProps & IWithForwardedRef
> = providedProps => {
  const {
    style,
    className,
    iconClassName,
    dense,
    forceSize,
    forceFontSize,
    children,
    forwardedRef,
    ...props
  } = providedProps as FontIconWithDefaultProps;

  return (
    <i
      {...props}
      ref={forwardedRef}
      style={createStyles(style, forceSize, forceFontSize, dense)}
      className={cn(
        "rmd-icon rmd-icon--font",
        {
          "rmd-icon--dense": dense,
        },
        iconClassName,
        className
      )}
    >
      {children}
    </i>
  );
};

const defaultProps: IFontIconDefaultProps = {
  dense: false,
  iconClassName: "material-icons",
  forceSize: false,
  forceFontSize: false,
};

FontIcon.defaultProps = defaultProps;

if (process.env.NODE_ENV !== "production") {
  let PropTypes = null;
  try {
    PropTypes = require("prop-types");
  } catch (e) {}

  if (PropTypes) {
    FontIcon.propTypes = {
      className: PropTypes.string,
      iconClassName: PropTypes.string,
      dense: PropTypes.bool,
      forceSize: PropTypes.oneOfType([PropTypes.bool, PropTypes.number]),
      forceFontSize: PropTypes.bool,
      children: PropTypes.node,
    };
  }
}

export default forwardRef<HTMLElement, IFontIconProps>((props, ref) => (
  <FontIcon {...props} forwardedRef={ref} />
));
