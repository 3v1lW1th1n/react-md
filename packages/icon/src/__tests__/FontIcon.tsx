import * as React from "react";
import { create } from "react-test-renderer";
import { mount } from "enzyme";

import FontIcon from "../FontIcon";

describe("FontIcon", () => {
  it("should use render as an <i> tag", () => {
    const icon = mount(<FontIcon>home</FontIcon>);

    expect(icon.find("i").length).toBe(1);
  });

  it("should render correctly", () => {
    expect(create(<FontIcon>home</FontIcon>).toJSON()).toMatchSnapshot();
    expect(
      create(<FontIcon iconClassName="fa fa-github" />).toJSON()
    ).toMatchSnapshot();
  });

  it("should apply the correct style prop", () => {
    const icon = mount(<FontIcon />);

    expect(icon.find("i").props().style).toBeUndefined();

    const style = { color: "red" };
    icon.setProps({ style });
    expect(icon.find("i").props().style).toEqual(style);

    icon.setProps({ style: undefined, forceSize: true });
    expect(icon.find("i").props().style).toEqual({
      height: 24,
      width: 24,
    });

    icon.setProps({ forceFontSize: true });
    expect(icon.find("i").props().style).toEqual({
      fontSize: 24,
      height: 24,
      width: 24,
    });

    icon.setProps({ forceSize: 100 });
    expect(icon.find("i").props().style).toEqual({
      fontSize: 100,
      height: 100,
      width: 100,
    });

    icon.setProps({ style });
    expect(icon.find("i").props().style).toEqual({
      fontSize: 100,
      height: 100,
      width: 100,
      color: style.color,
    });
  });

  it("should apply the correct class names", () => {
    const icon = mount(<FontIcon />);

    expect(icon.find("i").hasClass("rmd-icon")).toBe(true);
    expect(icon.find("i").hasClass("rmd-icon--font")).toBe(true);
    expect(icon.find("i").hasClass("rmd-icon--font-dense")).toBe(false);
    expect(icon.find("i").hasClass("material-icons")).toBe(true);
    expect(icon.render()).toMatchSnapshot();

    icon.setProps({ className: "test-thing", dense: true });
    expect(icon.find("i").hasClass("rmd-icon")).toBe(true);
    expect(icon.find("i").hasClass("rmd-icon--font")).toBe(true);
    expect(icon.find("i").hasClass("rmd-icon--font-dense")).toBe(true);
    expect(icon.find("i").hasClass("material-icons")).toBe(true);
    expect(icon.find("i").hasClass("test-thing")).toBe(true);
    expect(icon.render()).toMatchSnapshot();

    icon.setProps({ iconClassName: "fa fa-github" });
    expect(icon.find("i").hasClass("rmd-icon")).toBe(true);
    expect(icon.find("i").hasClass("rmd-icon--font")).toBe(true);
    expect(icon.find("i").hasClass("rmd-icon--font-dense")).toBe(true);
    expect(icon.find("i").hasClass("material-icons")).toBe(false);
    expect(icon.find("i").hasClass("test-thing")).toBe(true);
    expect(icon.find("i").hasClass("fa")).toBe(true);
    expect(icon.find("i").hasClass("fa-github")).toBe(true);
    expect(icon.render()).toMatchSnapshot();
  });
});
