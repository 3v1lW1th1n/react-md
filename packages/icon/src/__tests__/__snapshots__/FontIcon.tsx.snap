// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FontIcon should apply the correct class names 1`] = `
<i
  aria-hidden="true"
  class="rmd-icon rmd-icon--font material-icons"
/>
`;

exports[`FontIcon should apply the correct class names 2`] = `
<i
  aria-hidden="true"
  class="rmd-icon rmd-icon--font rmd-icon--dense material-icons test-thing"
/>
`;

exports[`FontIcon should apply the correct class names 3`] = `
<i
  aria-hidden="true"
  class="rmd-icon rmd-icon--font rmd-icon--dense fa fa-github test-thing"
/>
`;

exports[`FontIcon should render correctly 1`] = `
<i
  aria-hidden="true"
  className="rmd-icon rmd-icon--font material-icons"
>
  home
</i>
`;

exports[`FontIcon should render correctly 2`] = `
<i
  aria-hidden="true"
  className="rmd-icon rmd-icon--font fa fa-github"
/>
`;
